src/const-safety.md

```diff
@@ -0,0 +1,171 @@
+# Const safety
+
+The miri engine, which is used to execute code at compile time, can fail in
+four possible ways:
+
+* The program performs an unsupported operation (e.g., calling an unimplemented
+  intrinsics, or doing an operation that would observe the integer address of a
+  pointer).
+* The program causes undefined behavior (e.g., dereferencing an out-of-bounds
+  pointer).
+* The program panics (e.g., a failed bounds check).
+* The program exhausts its resources: It might overflow the stack, allocation
+  too much memory or loops forever.  Note that detecting these conditions
+  happens on a best-effort basis only.
+
+Just like panics and non-termination are acceptable in safe run-time Rust code,
+we also consider these acceptable in safe compile-time Rust code.  However, we
+would like to rule out the first two kinds of failures in safe code.  Following
+the terminology in [this blog post], we call a program that does not fail in the
+first two ways *const safe*.
+
+[this blog post]: https://www.ralfj.de/blog/2018/07/19/const.html
+
+The goal of the const safety check, then, is to ensure that a program is const
+safe.  What makes this tricky is that there are some operations that are safe as
+far as run-time Rust is concerned, but are inherently unsupportable in the miri engine and hence
+not const safe (they fall in the first category of failures above).
+We call these operations *unconst*.  The purpose
+of the following section is to explain this in more detail, before proceeding
+with the main definitions.
+
+## Miri background
+
+A very simple example of an unconst operation is
+
+```rust
+static S:i32 = 0;
+const BAD:bool = (&S as *const i32 as usize) % 16 == 0;
+```
+
+The modulo operation here is not supported by the miri engine because evaluating
+it requires knowing the actual integer address of `S`.
+
+The way miri handles this is by treating pointer and integer values separately.
+The most primitive kind of value in miri is a `Scalar`, and a scalar is *either*
+a pointer (`Scalar::Ptr`) or a bunch of bits representing an integer
+(`Scalar::Bits`).  Every value of a variable of primitive type is stored as a
```

Comment:
```suggestion
(`Scalar::Bits`), or uninitialized.  Every value of a variable of primitive type is stored as a
```
This is mentioned later so might be good to have that here for completeness' sake.

We should likely replace `Scalar` by just the work "scalar" then, and generally talk about concepts rather than specific rustc implementation types.